docs/src/references/agent/device-profile.md

```diff
@@ -0,0 +1,213 @@
+instead of multiple.
+
+**Q2: Why not model each device profile update as a firmware update that includes the desired software and configurations as well**
+
+Although this is a more robust approach, it is not feasible on all kinds of devices,
+especially the ones that does not support delta firmware updates.
+On such devices, pushing the entire firmware binary for each iterative change would considerably increase the binary size overhead.
+
+# Device profile capability
+
+A device that supports device profile operation must declare that capability by publishing the following message:
+
+```sh te2mqtt formats=v1
+tedge mqtt pub -r 'te/device/main///cmd/device_profile' '{
+  "name": "prod-profile",
+  "version: "v1"
+}'
+```
+
+If the current profile information is not known upfront, publish an empty JSON (`{}`) instead.
+The firmware information, supported software and configuration types must be declared separately with their respective capability messages.
+
+# Device profile command
+
+Once the `device_profile` capability is declared, the device can receive `device_profile` commands
+by subscribing to `<root>/<device-topic-id>/cmd/device_profile/+` MQTT topics.
+For example, subscribe to the following topic for the `main` device:
+
+```sh te2mqtt formats=v1
+tedge mqtt sub 'te/device/main///cmd/device_profile/+'
+```
+
+A `device_profile` command with `id` "1234" is triggered as follows:
+
+```sh te2mqtt formats=v1
+tedge mqtt pub -r 'te/device/main///cmd/device_profile/1234' '{
+  "status": "init",
+  "name": "prod-profile",
+  "version": "v2",
+  "profile": [
+    {
+      "module": "firmware",
+      "include": true,
+      "name": "core-image-tedge-rauc",
+      "remoteUrl": "https://abc.com/some/firmware/url",
+      "version": "20240430.1139"
+    },
+    {
+      "module": "software",
+      "include": true,
+      "name": "collectd",
+      "version": "latest",
+      "type": "apt",
+      "action": "install"
+    },
+    {
+      "module": "configuration",
+      "include": true,
+      "type": "collectd.conf",
+      "remoteUrl":"https://abc.com/some/collectd/conf",
+      "path": "/etc/collectd/collectd.conf"
+    },
+    {
+      "module": "software",
+      "include": true,
+      "name": "jq",
+      "version": "latest",
+      "type": "apt",
+      "action": "install"
+    }
+  ]
+}'
+```
+
+The profile definition in the payload is an array of modules.
+Each module could be a firmware, software or configuration.
+The modules are installed/applied in the order in which they are defined in the profile definition.
+There is no restriction on the order of modules in a profile and hence can be defined in any preferred order.
+For example, additional software can be installed or configurations updated before the firmware is updated.
+
+The `"include"` field is optional and the value is true, by default.
+It can be used to exclude already installed modules when the same profile is reapplied on a device after a partial failure.
```

Comment:
You mean by manually excluding modules that have been fully installed by a previous partial run?

This might be cumbersome. Couldn't this be a responsibility of the system?
E.g to persist a watermark of what has been done by the partial run, and to resume from there a new attempt.

Reply:
Yes, we can consider that. Since we've traditionally not stored/maintained any state about previously executed (esp. failed) operations, this was the only other alternative that I could think of, other than reapplying the same modules again and expect them to be idempotent.

Reply:
We should let each firmware/software/configuration handler be responsible for the idempotency. The device profile should just execute the same workflow again, and each individual handler can work out if it should really do something or not.

Reply:
>  The device profile should just execute the same workflow again, and each individual handler can work out if it should really do something or not.

This will also let the user fix one step or another.

Reply:
I prefer the new proposal to `skip` a step.